cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -335,4 +338,82 @@ int ztunnel_ingress(struct __sk_buff *skb)
     return TC_ACT_OK;
 }
 
+// For ztunnel pod veth pair(in pod ns) ingress hook
+// prog name disp limits to 15 bytes
+SEC("tc")
+int ztunnel_tproxy(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct bpf_sock_tuple *tuple;
+    size_t tuple_len;
+    struct bpf_sock *sk;
+    __u32 *log_level = get_log_level();
+    struct bpf_sock_tuple proxy_tup;
+    __be16 proxy_port;
+    int ret;
+
+    if (skb->cb[4] != OUTBOUND_CB && skb->cb[4] != INBOUND_CB)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support UDP tunneling
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+    tuple = (struct bpf_sock_tuple *)&iph->saddr;
+    tuple_len = sizeof(tuple->ipv4);
+    if ((void *)tuple + sizeof(tuple->ipv4) > (void *)(long)skb->data_end)
+        return TC_ACT_SHOT;
+
+    sk = bpf_skc_lookup_tcp(skb, tuple, tuple_len, BPF_F_CURRENT_NETNS, 0);
+    if (sk) {
+        if (sk->state != BPF_TCP_LISTEN) {
+            goto assign;
+        }
+        bpf_sk_release(sk);
+    }
```

Comment:
Can we structure the whole "if socket && state expected then use it, else twiddle tuple" logic so it's more obvious and doesn't use `goto`*?

*(I am not categorically or religiously opposed to `goto` but I don't think it helps readability here)

Reply:
updated.